NSFS - upload failed for 1 GB file with ENOSPC

### Environment info
- NooBaa Version: master-20210519 
- Platform: OCP 4.6.16

### Actual behavior
1. s3 upload of 1 GB file is not reflecting in s3://bucket

### Expected behavior
1.s3 upload should be successful and it should reflect the object in s3://bucket

### Steps to reproduce
1. Create namespace resource
2. Create account with UID, GID parameters
3. Create s3 bucket bucket-3
4. Create a 1 GB file
5. Run command to upload it to this bucket-3

Bucket status:
  Bucket                 : bucket-3
  Type                   : NAMESPACE
  Mode                   : OPTIMAL
  Num Objects            : 0
  Data Size              : 0.000 B
  Data Size Reduced      : 0.000 B
  Data Space Avail       : 0.000 B
  
File size:
```
[root@hpo-node1 akshat]# du -sh small_chunk_1
977M    small_chunk_1
```

Upload using command
`s3u3 cp small_chunk_1 s3://bucket-3/`

Upload failed:

 ` upload failed: ./small_chunk_1 to s3://bucket-3/small_chunk_1 Read timeout on endpoint URL: "https://api.ocp-akshat-1.cp.fyre.ibm.com:8443/bucket-3/small_chunk_1?uploadId=33f65ae4-9f2f-464a-bdaf-257ce0f2537a"`


### More information - Screenshots / Logs / Other output

```
sh-4.4# du -sh /nsfs/parent/bucket-3/.noobaa-nsfs/
977M	/nsfs/parent/bucket-3/.noobaa-nsfs/
```

```
[root@ocp-akshat-1-inf ~]# cat endpoint_small_chunk  |grep "PUT OBJECT PART bucket-3 small_chunk_1"|wc -l
123
```

```
May-20 20:06:54.646 [Endpoint/8] [ERROR] CONSOLE:: _upload_stream had error:  [Error: No space left on device] { code: 'ENOSPC' }
May-20 20:06:54.646 [Endpoint/8] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-1/YSWTisKg/25144.m3Od30T9Nkb%294y9r.rnd</Resource><RequestId>koxbrsts-f21tjf-zux</RequestId></Error> PUT /bucket-1/YSWTisKg/25144.m3Od30T9Nkb%294y9r.rnd {"user-agent":"MinIO (linux; amd64) minio-go/v7.0.11 warp/0.3.43","content-length":"1000","authorization":"AWS4-HMAC-SHA256 Credential=A83nz1ZeDJfi8ioJLAjT/20210520/us-east-1/s3/aws4_request, SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date, Signature=a37371d82a50a3824d1864397aea43eacbfec55fdbc298d8a77b0d6b6d4b4398","content-type":"application/octet-stream","x-amz-content-sha256":"UNSIGNED-PAYLOAD","x-amz-date":"20210520T200654Z","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: No space left on device
```
Seems like there is no space left on the device and this is why the upload fails.

Hello @jeniawhite , @guymguym 

I took latest PR from build 20210607.

Upload of 10GB file failed multiple times:

`s3u1 cp medium_chunk_2 s3://bucket-21/`

resulted in:

`upload failed: ./medium_chunk_2 to s3://bucket-21/medium_chunk_2 An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation: The specified key does not exist.
`
From endpoint log:

```
Jun-8 7:21:13.472 [Endpoint/9] [ERROR] CONSOLE::  [Error: No such file or directory] { code: 'ENOENT' }
Jun-8 7:21:13.476 [Endpoint/9] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resource>/bucket-21/medium_chunk_2?uploadId=bbf8bb2c-5268-4dab-8fc7-b3f193a440f4</Resource><RequestId>kpnprriv-4l61yc-19cc</RequestId></Error> POST /bucket-21/medium_chunk_2?uploadId=bbf8bb2c-5268-4dab-8fc7-b3f193a440f4 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.cp","x-amz-date":"20210608T072047Z","x-amz-content-sha256":"724c5e1855fb269e6b937c9869fc34c23fcf462938e73918aedbadd8d8563686","authorization":"AWS4-HMAC-SHA256 Credential=VdQhqra60t4VjbH1Z6gI/20210608/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=9c4396ab83da4c19c4db817f395a2153d37e394a7e468c41e738642d183c814c","content-length":"122810","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: No such file or directory
Jun-8 7:21:13.659 [Endpoint/9]    [L0] core.sdk.object_sdk:: _load_bucket_namespace {
  _id: '60bf0c44fb88b00022295013',
  name: SENSITIVE-4fb32e336f0d19ff,
  active_triggers: [ [length]: 0 ],
  system_owner: SENSITIVE-9cf0fd89409efef8,
  bucket_owner: SENSITIVE-3d288e97f8a66631,
  bucket_info: { name: SENSITIVE-4fb32e336f0d19ff, owner_account: { email: SENSITIVE-3d288e97f8a66631, id: '60b9eab981db7600272fde66' }, namespace: { write_resource: { resource: 'nsfs-nsr-1', path: '/bucket-21' }, read_resources: [ { resource: 'nsfs-nsr-1', path: '/bucket-21' }, [length]: 1 ], should_create_underlying_storage: true }, tag: '', num_objects: { value: 0, last_update: 1623133186130 }, writable: false, storage: { values: { used: 0, used_other: 0, total: 0, free: 0 }, last_update: 1623136755736 }, data: { size: 0, size_reduced: 0, free: 0, available_for_upload: 0, last_update: 1623136755736 }, usage_by_pool: { pools: [ [length]: 0 ], last_update: 1623133186130 }, stats_by_type: [ [length]: 0 ], mode: 'OPTIMAL', bucket_type: 'NAMESPACE', versioning: 'DISABLED', triggers: [ [length]: 0 ] },
  namespace: {
    write_resource: { resource: { fs_root_path: '/nsfs/nsfs-nsr-1', fs_backend: 'GPFS', id: '60b9e3f681db7600272fde65', name: 'nsfs-nsr-1' }, path: '/bucket-21' },
    read_resources: [ { resource: { fs_root_path: '/nsfs/nsfs-nsr-1', fs_backend: 'GPFS', id: '60b9e3f681db7600272fde65', name: 'nsfs-nsr-1' }, path: '/bucket-21' }, [length]: 1 ],
    should_create_underlying_storage: true
  }
}
Jun-8 7:21:13.660 [Endpoint/9]   [LOG] CONSOLE:: NamespaceFS: buffers_pool BuffersPool {
  buf_size: 8388608,
  buffers: [ <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 8388558 more bytes>, <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 8388558 more bytes> ],
  sem: Semaphore { _initial: 536870912, _value: 536870912, _waiting_value: 0, _wq: WaitQueue { _name: '_wq_', _q: <ref *1> LinkedList { next: '_lln_', prev: '_llp_', head: '_llh_', length: 0, _lln_: [Circular *1], _llp_: [Circular *1], _llh_: [Circular *1] } } }
}
Jun-8 7:21:13.662 [Endpoint/9] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchUpload</Code><Message>The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.</Message><Resource>/bucket-21/medium_chunk_2?uploadId=bbf8bb2c-5268-4dab-8fc7-b3f193a440f4</Resource><RequestId>kpnpsb6o-cjp60j-121c</RequestId></Error> DELETE /bucket-21/medium_chunk_2?uploadId=bbf8bb2c-5268-4dab-8fc7-b3f193a440f4 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.cp","x-amz-date":"20210608T072113Z","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","authorization":"AWS4-HMAC-SHA256 Credential=VdQhqra60t4VjbH1Z6gI/20210608/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=89b14475f17ec03cb08eb900528040bcf5e7a3789b8d8dfd646d6c7430296567","content-length":"0","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: No such file or directory
(END)

```

--- a/config.js
+++ b/config.js
@@ -106,6 +106,8 @@ config.N2N_OFFER_INTERNAL = false;
 
 config.AMZ_DATE_MAX_TIME_SKEW_MILLIS = 15 * 60 * 1000;
 config.ENDPOINT_MONITOR_INTERVAL = 10 * 60 * 1000; // 10min
+// Keep connection on long requests
+config.S3_KEEP_ALIVE_WHITESPACE_INTERVAL = 30 * 1000;
 
 ///////////////
 // MD CONFIG //

--- a/src/endpoint/s3/ops/s3_post_object_uploadId.js
+++ b/src/endpoint/s3/ops/s3_post_object_uploadId.js
@@ -24,6 +24,8 @@ async function post_object_uploadId(req, res) {
         throw new S3Error(S3Error.MalformedXML);
     }
 
+    http_utils.set_keep_alive_whitespace_interval(res);
+
     const reply = await req.object_sdk.complete_object_upload({
         obj_id: req.query.uploadId,
         bucket: req.params.bucket,
@@ -32,11 +34,15 @@ async function post_object_uploadId(req, res) {
         multiparts
     });
 
+    // TODO: Should be refactored when coding the complete solution
+    // We need to know all of these things prior to completion
+    // This also leaves a gap on long versioned/encrypted uploads (failure to assign headers after responding)
     s3_utils.set_encryption_response_headers(req, res, reply.encryption);
 
     if (reply.version_id && reply.version_id !== 'null') {
         res.setHeader('x-amz-version-id', reply.version_id);
     }
+
     return {
         CompleteMultipartUploadResult: {
             Bucket: req.params.bucket,

--- a/src/endpoint/s3/s3_errors.js
+++ b/src/endpoint/s3/s3_errors.js
@@ -502,4 +502,38 @@ S3Error.ObjectLockConfigurationNotFoundError = Object.freeze({
     http_code: 404,
 });
 
+S3Error.RPC_ERRORS_TO_S3 = Object.freeze({
+    UNAUTHORIZED: S3Error.AccessDenied,
+    BAD_REQUEST: S3Error.BadRequest,
+    FORBIDDEN: S3Error.AccessDenied,
+    NO_SUCH_BUCKET: S3Error.NoSuchBucket,
+    NO_SUCH_OBJECT: S3Error.NoSuchKey,
+    INVALID_BUCKET_NAME: S3Error.InvalidBucketName,
+    NOT_EMPTY: S3Error.BucketNotEmpty,
+    BUCKET_ALREADY_EXISTS: S3Error.BucketAlreadyExists,
+    BUCKET_ALREADY_OWNED_BY_YOU: S3Error.BucketAlreadyOwnedByYou,
+    NO_SUCH_UPLOAD: S3Error.NoSuchUpload,
+    BAD_DIGEST_MD5: S3Error.BadDigest,
+    BAD_DIGEST_SHA256: S3Error.XAmzContentSHA256Mismatch,
+    BAD_SIZE: S3Error.IncompleteBody,
+    IF_MODIFIED_SINCE: S3Error.NotModified,
+    IF_UNMODIFIED_SINCE: S3Error.PreconditionFailed,
+    IF_MATCH_ETAG: S3Error.PreconditionFailed,
+    IF_NONE_MATCH_ETAG: S3Error.NotModified,
+    OBJECT_IO_STREAM_ITEM_TIMEOUT: S3Error.SlowDown,
+    INVALID_PART: S3Error.InvalidPart,
+    INVALID_PORT_ORDER: S3Error.InvalidPartOrder,
+    INVALID_BUCKET_STATE: S3Error.InvalidBucketState,
+    NOT_ENOUGH_SPACE: S3Error.InvalidBucketState,
+    MALFORMED_POLICY: S3Error.MalformedPolicy,
+    NO_SUCH_OBJECT_LOCK_CONFIGURATION: S3Error.NoSuchObjectLockConfiguration,
+    OBJECT_LOCK_CONFIGURATION_NOT_FOUND_ERROR: S3Error.ObjectLockConfigurationNotFoundError,
+    INVALID_REQUEST: S3Error.InvalidRequest,
+    NOT_IMPLEMENTED: S3Error.NotImplemented,
+    INVALID_ACCESS_KEY_ID: S3Error.InvalidAccessKeyId,
+    SIGNATURE_DOES_NOT_MATCH: S3Error.SignatureDoesNotMatch,
+    SERVICE_UNAVAILABLE: S3Error.ServiceUnavailable,
+    INVALID_RANGE: S3Error.InvalidRange,
+});
+
 exports.S3Error = S3Error;

--- a/src/endpoint/s3/s3_rest.js
+++ b/src/endpoint/s3/s3_rest.js
@@ -59,40 +59,6 @@ const OBJECT_SUB_RESOURCES = Object.freeze({
 const UNSIGNED_PAYLOAD = 'UNSIGNED-PAYLOAD';
 const STREAMING_PAYLOAD = 'STREAMING-AWS4-HMAC-SHA256-PAYLOAD';
 
-const RPC_ERRORS_TO_S3 = Object.freeze({
-    UNAUTHORIZED: S3Error.AccessDenied,
-    BAD_REQUEST: S3Error.BadRequest,
-    FORBIDDEN: S3Error.AccessDenied,
-    NO_SUCH_BUCKET: S3Error.NoSuchBucket,
-    NO_SUCH_OBJECT: S3Error.NoSuchKey,
-    INVALID_BUCKET_NAME: S3Error.InvalidBucketName,
-    NOT_EMPTY: S3Error.BucketNotEmpty,
-    BUCKET_ALREADY_EXISTS: S3Error.BucketAlreadyExists,
-    BUCKET_ALREADY_OWNED_BY_YOU: S3Error.BucketAlreadyOwnedByYou,
-    NO_SUCH_UPLOAD: S3Error.NoSuchUpload,
-    BAD_DIGEST_MD5: S3Error.BadDigest,
-    BAD_DIGEST_SHA256: S3Error.XAmzContentSHA256Mismatch,
-    BAD_SIZE: S3Error.IncompleteBody,
-    IF_MODIFIED_SINCE: S3Error.NotModified,
-    IF_UNMODIFIED_SINCE: S3Error.PreconditionFailed,
-    IF_MATCH_ETAG: S3Error.PreconditionFailed,
-    IF_NONE_MATCH_ETAG: S3Error.NotModified,
-    OBJECT_IO_STREAM_ITEM_TIMEOUT: S3Error.SlowDown,
-    INVALID_PART: S3Error.InvalidPart,
-    INVALID_PORT_ORDER: S3Error.InvalidPartOrder,
-    INVALID_BUCKET_STATE: S3Error.InvalidBucketState,
-    NOT_ENOUGH_SPACE: S3Error.InvalidBucketState,
-    MALFORMED_POLICY: S3Error.MalformedPolicy,
-    NO_SUCH_OBJECT_LOCK_CONFIGURATION: S3Error.NoSuchObjectLockConfiguration,
-    OBJECT_LOCK_CONFIGURATION_NOT_FOUND_ERROR: S3Error.ObjectLockConfigurationNotFoundError,
-    INVALID_REQUEST: S3Error.InvalidRequest,
-    NOT_IMPLEMENTED: S3Error.NotImplemented,
-    INVALID_ACCESS_KEY_ID: S3Error.InvalidAccessKeyId,
-    SIGNATURE_DOES_NOT_MATCH: S3Error.SignatureDoesNotMatch,
-    SERVICE_UNAVAILABLE: S3Error.ServiceUnavailable,
-    INVALID_RANGE: S3Error.InvalidRange,
-});
-
 const S3_OPS = load_ops();
 
 let usage_report = new_usage_report();
@@ -446,7 +412,7 @@ function parse_op_name(req) {
 function handle_error(req, res, err) {
     var s3err =
         ((err instanceof S3Error) && err) ||
-        new S3Error(RPC_ERRORS_TO_S3[err.rpc_code] || S3Error.InternalError);
+        new S3Error(S3Error.RPC_ERRORS_TO_S3[err.rpc_code] || S3Error.InternalError);
 
     if (s3err.code === 'MalformedPolicy') {
         s3err.message = err.message;
@@ -455,10 +421,18 @@ function handle_error(req, res, err) {
 
     if (s3err.rpc_data) {
         if (s3err.rpc_data.etag) {
-            res.setHeader('ETag', s3err.rpc_data.etag);
+            if (res.headersSent) {
+                dbg.log0('Sent reply in body, bit too late for Etag header');
+            } else {
+                res.setHeader('ETag', s3err.rpc_data.etag);
+            }
         }
         if (s3err.rpc_data.last_modified) {
-            res.setHeader('Last-Modified', time_utils.format_http_header_date(new Date(s3err.rpc_data.last_modified)));
+            if (res.headersSent) {
+                dbg.log0('Sent reply in body, bit too late for Last-Modified header');
+            } else {
+                res.setHeader('Last-Modified', time_utils.format_http_header_date(new Date(s3err.rpc_data.last_modified)));
+            }
         }
     }
 
@@ -475,16 +449,20 @@ function handle_error(req, res, err) {
         req.method, req.originalUrl,
         JSON.stringify(req.headers),
         err.stack || err);
-    res.statusCode = s3err.http_code;
-    res.setHeader('Content-Type', 'application/xml');
-    res.setHeader('Content-Length', Buffer.byteLength(reply));
+    if (res.headersSent) {
+        dbg.log0('Sending error xml in body, but too late for headers...');
+    } else {
+        res.statusCode = s3err.http_code;
+        res.setHeader('Content-Type', 'application/xml');
+        res.setHeader('Content-Length', Buffer.byteLength(reply));
+    }
     res.end(reply);
 }
 
 async function _handle_html_response(req, res, err) {
     var s3err =
         ((err instanceof S3Error) && err) ||
-        new S3Error(RPC_ERRORS_TO_S3[err.rpc_code] || S3Error.InternalError);
+        new S3Error(S3Error.RPC_ERRORS_TO_S3[err.rpc_code] || S3Error.InternalError);
 
     if (s3err.rpc_data) {
         if (s3err.rpc_data.etag) {

--- a/src/util/http_utils.js
+++ b/src/util/http_utils.js
@@ -15,6 +15,7 @@ const createHttpsProxyAgent = require('https-proxy-agent');
 const dbg = require('./debug_module')(__filename);
 const xml_utils = require('./xml_utils');
 const cloud_utils = require('./cloud_utils');
+const config = require('../../config');
 
 const { HTTP_PROXY, HTTPS_PROXY, NO_PROXY } = process.env;
 const http_agent = new http.Agent();
@@ -321,13 +322,19 @@ function send_reply(req, res, reply, options) {
                 }
             }) :
             reply;
-        const xml_reply = xml_utils.encode_xml(xml_root);
+        // TODO: Refactor later on to support potential headers response and delayed XML body
+        // This is done for the complete since we assign the XML header only in body prior to responding
+        const xml_reply = xml_utils.encode_xml(xml_root, /* ignore_header */ res.headersSent);
         dbg.log1('HTTP REPLY XML', req.method, req.originalUrl,
             JSON.stringify(req.headers),
             xml_reply.length <= 2000 ?
             xml_reply : xml_reply.slice(0, 1000) + ' ... ' + xml_reply.slice(-1000));
-        res.setHeader('Content-Type', 'application/xml');
-        res.setHeader('Content-Length', Buffer.byteLength(xml_reply));
+        if (res.headersSent) {
+            dbg.log0('Sending xml reply in body, bit too late for headers');
+        } else {
+            res.setHeader('Content-Type', 'application/xml');
+            res.setHeader('Content-Length', Buffer.byteLength(xml_reply));
+        }
         res.end(xml_reply);
         return;
     }
@@ -444,6 +451,30 @@ function make_https_request(options, body, body_encoding) {
     });
 }
 
+// Write periodically to keep the connection alive
+// TODO: Every complete above the S3_KEEP_ALIVE_WHITESPACE_INTERVAL
+// will be responded in BODY, we cannot rollback from that for pre HEADER failures
+// instread of having a fine tuned way start of sending body to make sure that we won't respond by BODY on HEADER failures
+function set_keep_alive_whitespace_interval(res) {
+    let count = 0;
+    res.keep_alive_whitespace_interval = setInterval(() => {
+        count += 1;
+        dbg.log0(`keep_alive_whitespace_interval headersSent=${res.headersSent} count=${count}`);
+        if (res.headersSent) {
+            // Keep the connection alive with white spaces
+            res.write(' ');
+        } else {
+            // Mark begining of BODY response
+            // Assuming that write automatically assigns 200 OK status
+            res.write(xml_utils.XML_HEADER);
+        }
+    }, config.S3_KEEP_ALIVE_WHITESPACE_INTERVAL);
+    res.keep_alive_whitespace_interval.unref();
+    const clear = () => clearInterval(res.keep_alive_whitespace_interval);
+    res.on('close', clear);
+    res.on('finish', clear);
+}
+
 exports.parse_url_query = parse_url_query;
 exports.parse_client_ip = parse_client_ip;
 exports.get_md_conditions = get_md_conditions;
@@ -459,4 +490,5 @@ exports.get_unsecured_agent = get_unsecured_agent;
 exports.update_http_agents = update_http_agents;
 exports.update_https_agents = update_https_agents;
 exports.make_https_request = make_https_request;
+exports.set_keep_alive_whitespace_interval = set_keep_alive_whitespace_interval;
 exports.parse_xml_to_js = parse_xml_to_js;

--- a/src/util/xml_utils.js
+++ b/src/util/xml_utils.js
@@ -1,8 +1,16 @@
 /* Copyright (C) 2016 NooBaa */
 'use strict';
 
-exports.encode_xml = encode_xml;
-exports.encode_xml_str = encode_xml_str;
+// see https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Predefined_entities_in_XML
+const XML_CHAR_ENTITY_MAP = Object.freeze({
+    '"': '&quot;',
+    '&': '&amp;',
+    '\'': '&apos;',
+    '<': '&lt;',
+    '>': '&gt;'
+});
+
+const XML_HEADER = '<?xml version="1.0" encoding="UTF-8"?>';
 
 /**
  *
@@ -23,8 +31,8 @@ exports.encode_xml_str = encode_xml_str;
  *   <root> <a>1</a> <b>2</b> <a>3</a> <b>4</b> <z>42</z> <c d="5"></c> </root>
  *
  */
-function encode_xml(object) {
-    let output = '<?xml version="1.0" encoding="UTF-8"?>';
+function encode_xml(object, ignore_header) {
+    let output = ignore_header ? '' : XML_HEADER;
     append_object(s => {
         output += s;
     }, object);
@@ -82,15 +90,10 @@ function append_object(append, object) {
     }
 }
 
-// see https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Predefined_entities_in_XML
-const XML_CHAR_ENTITY_MAP = Object.freeze({
-    '"': '&quot;',
-    '&': '&amp;',
-    '\'': '&apos;',
-    '<': '&lt;',
-    '>': '&gt;'
-});
-
 function encode_xml_str(s) {
     return String(s).replace(/(["&'<>])/g, (str, ch) => XML_CHAR_ENTITY_MAP[ch]);
 }
+
+exports.encode_xml = encode_xml;
+exports.encode_xml_str = encode_xml_str;
+exports.XML_HEADER = XML_HEADER;

